File: Firmware/src/StateMachine.py
import time
from enum import Enum

# State definitions
class State(Enum):
    IDLE = 0
    START = 1
    HIGH = 2
    LOW = 3
    SHUTDOWN = 4

class StageHandler:
    def __init__(self, stages, next_state):
        """
        Initialize the stage handler.

        :param stages: List of dictionaries with 'function' and 'duration' or 'condition' keys.
        :param next_state: The state to transition to after all stages are complete.
        """
        self.stages = stages
        self.next_state = next_state
        self.current_stage_index = 0
        self.stage_start_time = None

    def tick(self):
        """Handle the execution of the current stage."""
        current_stage = self.stages[self.current_stage_index]

        # Check if the stage is starting for the first time
        if self.stage_start_time is None:
            self.stage_start_time = time.time()
            current_stage["function"]()

        # Handle stages with a fixed duration
        if "duration" in current_stage:
            elapsed_time = time.time() - self.stage_start_time
            if elapsed_time >= current_stage["duration"]:
                self._advance_stage()

        # Handle stages with a condition for completion
        if "condition" in current_stage and current_stage["condition"]():
            self._advance_stage()

        # Check if all stages are complete
        if self.current_stage_index >= len(self.stages):
            return self.next_state  # Return the next state to transition to

        return None  # No transition yet, continue processing current stages

    def _advance_stage(self):
        """Advance to the next stage and reset the stage start time."""
        self.current_stage_index += 1
        self.stage_start_time = None


class StateMachine:
    def __init__(self):
        self.current_state = State.IDLE
        self.control_signals = {
            "RUN": False,
            "WATER_TEMP": 20.0
        }
        self.handlers = {}

        self.state_actions = {
            State.IDLE: lambda: print("System is IDLE. Waiting for RUN to become True."),
            State.START: self.run_stage_handler,
            State.HIGH: self.run_stage_handler,
            State.LOW: self.run_stage_handler,
            State.SHUTDOWN: self.run_stage_handler
        }

    def init_start_handler(self):
        return StageHandler(
            stages=[
                {"function": lambda: print("START: Stage 1 - Initializing component A."), "duration": 2.0},
                {"function": lambda: print("START: Stage 2 - Initializing component B."), "duration": 2.0},
                {"function": lambda: print("START: Stage 3 - Initializing component C."), "duration": 2.0},
                {"function": lambda: print("START: Stage 4 - Power check."), "duration": 2.0},
                {"function": lambda: print("START: Stage 5 - System ready."), "duration": 2.0}
            ],
            next_state=State.HIGH
        )

    def init_high_handler(self):
        return StageHandler(
            stages=[
                {"function": lambda: print(f"HIGH: Stage 1 - Increasing power. Temp: {self.control_signals['WATER_TEMP']:.1f}°C"), "duration": 5.0},
                {"function": lambda: print(f"HIGH: Stage 2 - Monitoring temperature. Temp: {self.control_signals['WATER_TEMP']:.1f}°C"),
                 "condition": lambda: self.control_signals["WATER_TEMP"] > 65}
            ],
            next_state=State.LOW
        )

    def init_low_handler(self):
        return StageHandler(
            stages=[
                {"function": lambda: print(f"LOW: Stage 1 - Reducing power. Temp: {self.control_signals['WATER_TEMP']:.1f}°C"), "duration": 5.0},
                {"function": lambda: print(f"LOW: Stage 2 - Monitoring temperature. Temp: {self.control_signals['WATER_TEMP']:.1f}°C"),
                 "condition": lambda: self.control_signals["WATER_TEMP"] < 55}
            ],
            next_state=State.HIGH
        )

    def init_shutdown_handler(self):
        return StageHandler(
            stages=[
                {"function": lambda: print("SHUTDOWN: Stage 1 - Initializing shutdown sequence."), "duration": 5.0},
                {"function": lambda: print("SHUTDOWN: Stage 2 - Finalizing shutdown sequence."), "duration": 10.0}
            ],
            next_state=State.IDLE
        )

    def run_stage_handler(self):
        """Run the current stage handler."""
        # Check for immediate transition to SHUTDOWN if RUN goes low
        if not self.control_signals["RUN"] and self.current_state != State.SHUTDOWN:
            print("RUN is False. Transitioning to SHUTDOWN.")
            self.current_state = State.SHUTDOWN
            self.handlers[self.current_state] = self.init_shutdown_handler()
            return


        if self.current_state not in self.handlers:
            # Initialize the handler for this state
            if self.current_state == State.HIGH:
                self.handlers[self.current_state] = self.init_high_handler()
            elif self.current_state == State.LOW:
                self.handlers[self.current_state] = self.init_low_handler()
            elif self.current_state == State.START:
                self.handlers[self.current_state] = self.init_start_handler()
            elif self.current_state == State.SHUTDOWN:
                self.handlers[self.current_state] = self.init_shutdown_handler()

        # Run the handler
        next_state = self.handlers[self.current_state].tick()
        if next_state:
            print(f"{self.current_state.name} complete. Transitioning to {next_state.name}.")
            self.current_state = next_state

    def idle_transitions(self):
        if self.control_signals["RUN"]:
            return State.START
        return State.IDLE

    def run(self):
        """Main loop for the state machine."""
        state_transitions = {
            State.IDLE: self.idle_transitions,
            State.START: lambda: self.current_state,  # Controlled by handler
            State.HIGH: lambda: self.current_state,   # Controlled by handler
            State.LOW: lambda: self.current_state,    # Controlled by handler
            State.SHUTDOWN: lambda: self.current_state  # Controlled by handler
        }

        while True:
            # Perform the current state's action
            self.state_actions[self.current_state]()

            # Determine the next state
            if self.current_state == State.IDLE:
                next_state = state_transitions[self.current_state]()
                if next_state != self.current_state:
                    print(f"Transitioning from {self.current_state.name} to {next_state.name}")
                    self.current_state = next_state

            # Simulate timing
            time.sleep(0.5)

File: Firmware/src/StateMachine.cpp
#include "StateMachine.h"
#include "HardwareInterface.h"
#include "Stages.h"

// Global pointer for condition lambdas
static StateMachine* smInstance = nullptr;

State StateMachine::largeCondition() {
    if (smInstance) {
        int waterTemp = smInstance->hardware.getWaterTemp();

        if (waterTemp > OVERTEMP_THRESHOLD) {
            return State::SHUTDOWN;
        } else if (waterTemp > LARGE_TO_SMALL_THRESHOLD) {
            return State::SMALL;
        }
    }
    return State::LARGE;
}


State StateMachine::smallCondition() {
    if (smInstance) {
        int waterTemp = smInstance->hardware.getWaterTemp();

        if (waterTemp > OVERTEMP_THRESHOLD) {
            return State::SHUTDOWN;
        } else if (waterTemp < SMALL_TO_LARGE_THRESHOLD) {
            return State::LARGE;
        }
    }
    return State::SMALL;
}

// Linear interpolation helper
int linearInterpolate(int start, int end, unsigned long elapsedTime, unsigned long duration) {
    float progress = static_cast<float>(elapsedTime) / (duration * 1000);
    if (progress > 1.0) progress = 1.0; // Clamp progress
    return start + static_cast<int>((end - start) * progress);
}

// Constructor
StateMachine::StateMachine(HardwareInterface& hw) : hardware(hw), currentState(IDLE), stageStartTime(0), currentStageIndex(0),
                                                    currentStages(nullptr), totalStages(0), nextState(IDLE),
                                                    runSignal(false) {
    smInstance = this;
}

// Initialize the state machine
void StateMachine::init() {
    Serial.begin(9600);
    resetHandler(IDLE);
}

// Set the RUN signal
void StateMachine::setRunSignal(bool run) {
    if (runSignal != run) {
        runSignal = run;
        if (!runSignal && currentState != SHUTDOWN) {
            currentState = SHUTDOWN;
            resetHandler(SHUTDOWN);
        }
    }
}

// Tick the state machine
void StateMachine::tick() {
    if (currentState == IDLE && runSignal) {
        currentState = START;
        resetHandler(START);
    }

    tickHandler();
}

// Reset the stage handler for a given state
void StateMachine::resetHandler(State state) {
    switch (state) {
        case SHUTDOWN:
            currentStages = shutdownStages;
            totalStages = shutdownStagesCount;
            nextState = IDLE;
            break;
        case START:
            currentStages = startStages;
            totalStages = startStagesCount;
            nextState = LARGE; // Transition to LARGE after START
            break;
        case LARGE: // Renamed from HIGH
            currentStages = largeStages;
            totalStages = largeStagesCount;
            nextState = SMALL; // Transition to SMALL
            break;
        case SMALL: // Renamed from LOW
            currentStages = smallStages;
            totalStages = smallStagesCount;
            nextState = LARGE; // Transition to LARGE
            break;
        default:
            currentStages = nullptr;
            totalStages = 0;
            break;
    }
    currentStageIndex = 0;
    stageStartTime = millis();
}

void StateMachine::tickHandler() {
    if (currentStageIndex >= totalStages) {
        // Transition to the next state
        currentState = nextState;
        resetHandler(currentState);
        return;
    }

    Stage& currentStage = currentStages[currentStageIndex];

    // Log the stage message
    Serial.println(currentStage.message);

    // Set digital states via hardware abstraction
    hardware.setWaterPumpState(currentStage.waterPumpState);
    hardware.setBlowerState(currentStage.blowerState);
    hardware.setGlowState(currentStage.glowState);

    // Calculate elapsed time
    unsigned long elapsedTime = millis() - stageStartTime;

    // Interpolate analog values
    int interpolatedFanSpeed = linearInterpolate(currentStage.fanSpeed.start, currentStage.fanSpeed.end, elapsedTime, currentStage.duration);
    int interpolatedFuelPump = linearInterpolate(currentStage.fuelPump.start, currentStage.fuelPump.end, elapsedTime, currentStage.duration);

    // Write interpolated values to hardware
    hardware.setFanSpeed(interpolatedFanSpeed);
    hardware.setFuelPumpSpeed(interpolatedFuelPump);

    // Log interpolated values
    Serial.print("Fan Speed: ");
    Serial.println(interpolatedFanSpeed);
    Serial.print("Fuel Pump: ");
    Serial.println(interpolatedFuelPump);

    // Check if the stage is complete
    if (elapsedTime >= currentStage.duration * 1000) {
        currentStageIndex++;
        stageStartTime = millis();

        // Final stage logic
        if (currentStageIndex >= totalStages) {
            State transitionState = currentStage.condition ? currentStage.condition() : nextState;

            // Handle state transition
            if (transitionState != currentState) {
                Serial.println("Transitioning to next state.");
                currentState = transitionState;
                resetHandler(currentState);
            }
        }
    }
}
